File: common/src/main/java/net/impleri/playerskills/api/Skill.java
package net.impleri.playerskills.api;

import net.minecraft.ChatFormatting;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * Base generic skill. These are meant to be nothing more than
 * containers for data. All logic should be handled by the SkillType.
 */
public class Skill<T> {
    public static final int UNLIMITED_CHANGES = -1;
    
    protected ResourceLocation name;
    protected ResourceLocation type;
    protected boolean notify;
    protected String notifyKey;
    protected TeamMode teamMode;
    protected T value;
    protected String description;
    protected List<T> options;
    protected int changesAllowed;

    public Skill(ResourceLocation name, ResourceLocation type, T value, String description, List<T> options, int changesAllowed, TeamMode teamMode, boolean notify, String notifyKey) {
        this.name = name;
        this.type = type;
        this.value = value;
        this.description = description;
        this.options = options;
        this.changesAllowed = changesAllowed;
        this.teamMode = teamMode;
        this.notify = notify;
        this.notifyKey = notifyKey;
    }

    public Skill<T> copy() {
        return copy(value, changesAllowed);
    }

    public Skill<T> copy(T value, int changesAllowed) {
        return new Skill<T>(name, type, value, description, options, changesAllowed, teamMode, notify, notifyKey);
    }

    public ResourceLocation getName() {
        return name;
    }

    public void setName(ResourceLocation name) {
        this.name = name;
    }

    public ResourceLocation getType() {
        return type;
    }

    public void setType(ResourceLocation type) {
        this.type = type;
    }

    public TeamMode getTeamMode() {
        return teamMode;
    }

    public void setTeamMode(TeamMode teamMode) {
        this.teamMode = teamMode;
    }

    @Nullable
    public T getValue() {
        return value;
    }

    public void setValue(@Nullable T value) {
        this.value = value;
    }

    @Nullable
    public String getDescription() {
        return description;
    }

    public void setDescription(@Nullable String description) {
        this.description = description;
    }

    @NotNull
    public List<T> getOptions() {
        return options;
    }

    public void setOptions(@NotNull List<T> options) {
        this.options = options;
    }

    public int getChangesAllowed() {
        return changesAllowed;
    }

    public boolean areChangesAllowed() {
        return changesAllowed != 0;
    }

    public void consumeChange() {
        if (changesAllowed > 0) {
            changesAllowed--;
        }
    }

    @Nullable
    public Component getNotification(T oldValue) {
        if (!notify || value == null) {
            return null;
        }

        return notifyKey == null ? getDefaultNotification() : formatNotification(notifyKey, oldValue);
    }

    protected Component getDefaultNotification() {
        return formatNotification("playerskills.notify.skill_change");
    }

    protected Component formatNotification(String messageKey) {
        return formatNotification(messageKey, null);
    }

    protected Component formatNotification(String messageKey, @Nullable T oldValue) {
        var skillName = formatSkillName();
        var skillValue = formatSkillValue();
        var oldSkillValue = oldValue == null ? "" : formatSkillValue(oldValue);

        return Component.translatable(messageKey, skillName, skillValue, oldSkillValue);
    }

    protected Component formatSkillName() {
        return Component.literal(getName().getPath().replace("_", " "))
                .withStyle(ChatFormatting.DARK_AQUA)
                .withStyle(ChatFormatting.BOLD);
    }

    protected Component formatSkillValue() {
        return formatSkillValue(value);
    }

    protected Component formatSkillValue(T value) {
        return Component.literal(value.toString())
                .withStyle(ChatFormatting.GOLD);
    }


    public boolean isAllowedValue(T nextVal) {
        return options.size() == 0 || options.contains(nextVal);
    }

    public boolean isSameAs(Skill<?> that) {
        return that.name == this.name;
    }

    public boolean isSameType(Skill<?> that) {
        return that.type == this.type;
    }

    private <V> boolean canEquals(Skill<V> that) {
        return that.getClass().isInstance(this);
    }

    @Override
    public boolean equals(Object that) {
        if (that instanceof Skill<?> skill) {
            return skill.canEquals(this) && isSameAs(skill) && isSameType(skill);
        }

        return false;
    }

    @Override
    public int hashCode() {
        return name.hashCode() * type.hashCode();
    }
}

File: common/src/main/kotlin/net/impleri/playerskills/api/Skill.kt
package net.impleri.playerskills.api

import net.impleri.playerskills.PlayerSkills
import net.impleri.playerskills.skills.registry.RegistryItemNotFound
import net.impleri.playerskills.skills.registry.Skills
import net.impleri.playerskills.utils.SkillResourceLocation
import net.minecraft.ChatFormatting
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation

/**
 * Base generic skill. These are meant to be nothing more than
 * containers for data. All logic should be handled by the SkillType.
 */
abstract class Skill<T>(
  val name: ResourceLocation,
  val type: ResourceLocation,
  val value: T? = null,
  val description: String? = null,
  val options: List<T> = ArrayList(),
  val changesAllowed: Int = UNLIMITED_CHANGES,
  val teamMode: TeamMode = TeamMode.off(),
  val notify: Boolean = false,
  val notifyKey: String? = null,
) {
  abstract fun copy(value: T?, changesAllowed: Int): Skill<T>

  protected open fun getDefaultNotification(): Component {
    return formatNotification("playerskills.notify.skill_change")
  }

  fun areChangesAllowed(): Boolean {
    return changesAllowed != 0
  }

  fun copy(): Skill<T> {
    return copy(value, changesAllowed)
  }

  fun change(newValue: T?): Skill<T> {
    return if (areChangesAllowed()) copy(newValue, changesAllowed - 1) else copy()
  }

  @JvmOverloads
  fun getNotification(oldValue: T? = null): Component? {
    if (!notify || value == null) {
      return null
    }

    return if (notifyKey == null) getDefaultNotification() else formatNotification(notifyKey, oldValue)
  }

  @JvmOverloads
  protected fun formatNotification(messageKey: String, oldValue: T? = null): Component {
    val skillName = formatSkillName()
    val skillValue = formatSkillValue()
    val oldSkillValue = if (oldValue == null) Component.literal("") else formatSkillValue(oldValue)

    return Component.translatable(messageKey, skillName, skillValue, oldSkillValue)
  }

  protected fun formatSkillName(): Component {
    return Component.literal(name.path.replace("_", " "))
      .withStyle(ChatFormatting.DARK_AQUA)
      .withStyle(ChatFormatting.BOLD)
  }

  @JvmOverloads
  protected fun formatSkillValue(value: T? = null): Component {
    return Component.literal(value.toString())
      .withStyle(ChatFormatting.GOLD)
  }

  fun isAllowedValue(nextVal: T?): Boolean {
    return nextVal == null || options.isEmpty() || options.contains(nextVal)
  }

  @Suppress("UNCHECKED_CAST")
  fun <V> cast(): Skill<V> {
    return this as Skill<V>
  }

  private fun isSameAs(that: Skill<*>): Boolean {
    return that.name == this.name
  }

  private fun isSameType(that: Skill<*>): Boolean {
    return that.type == this.type
  }

  private fun <V> canEquals(that: Skill<V>): Boolean {
    return that::class.isInstance(this)
  }

  override fun equals(other: Any?): Boolean {
    if (other is Skill<*>) {
      return other.canEquals(this) && isSameAs(other) && isSameType(other)
    }

    return false
  }

  override fun hashCode(): Int {
    return name.hashCode() * type.hashCode()
  }

  companion object {
    const val UNLIMITED_CHANGES = -1

    @JvmField
    val REGISTRY_KEY = Skills.REGISTRY_KEY

    fun all(): List<Skill<*>> {
      return Skills.entries()
    }

    fun <T> modify(skill: Skill<T>): Boolean {
      PlayerSkills.LOGGER.info("Saving skill ${skill.name}")
      return Skills.upsert(skill)
    }

    fun <T> remove(skill: Skill<T>): Boolean {
      return Skills.remove(skill)
    }

    /**
     * Find a Skill by string
     */
    @Throws(RegistryItemNotFound::class)
    fun <V> findOrThrow(name: String): Skill<V> {
      return findOrThrow(SkillResourceLocation.of(name))
    }

    /**
     * Find a Skill by name
     */
    @Throws(RegistryItemNotFound::class)
    fun <V> findOrThrow(location: ResourceLocation): Skill<V> {
      return Skills.findOrThrow(location)
    }

    fun <V> find(location: ResourceLocation): Skill<V>? {
      return Skills.find(location)
    }

    fun <V> find(location: String): Skill<V>? {
      return find(SkillResourceLocation.of(location))
    }
  }
}
